install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -16,3 +16,49 @@ spec:
   selector:
     matchLabels:
       k8s-app: cluster-version-operator
+---
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    k8s-app: cluster-version-operator
+  name: cluster-version-operator
+  namespace: openshift-cluster-version
+spec:
+  groups:
+  - name: cluster-version
+    rules:
+    - alert: ClusterVersionOperatorDown
+      annotations:
+        message: Cluster version operator has disappeared from Prometheus target discovery. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        absent(up{job="cluster-version-operator"} == 1)
+      for: 15m
+      labels:
+        severity: critical
+  - name: cluster-operators
+    rules:
+    - alert: ClusterOperatorDown
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has not been available for 10 mins. Operator may be down or disabled, cluster will not be kept up to date and upgrades will not be possible.
+      expr: |
+        cluster_operator_up{job="cluster-version-operator"} == 0
+      for: 10m
+      labels:
+        severity: critical
+    - alert: ClusterOperatorDegraded
+      annotations:
+        message: Cluster operator {{ "{{ $labels.name }}" }} has been degraded for 10 mins. Operator is degraded because {{ "{{ $labels.reason }}" }} and cluster upgrades will be unstable.
+      expr: |
+        cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"} == 1
+      for: 10m
+      labels:
+        severity: critical
```

Comment:
I understand the intend of this, but a cluster administrator is not going to know what to do here. This goes back to my general critizism about using `cluster_operator_conditions` for any type of alerting. Apart from the indirections it imposes, this is the wrong approach to monitoring. Instead we should be setting SLOs for each component, and alert on the failure to deliver those. Those alerts can then have clear runbooks as to what actionable things can be done by the cluster administrator about this particular component.

I think we can have alerts that go in a similar direction as thi this on the _telemeter_ side, so our engineering teams can act on this, but I don't think this is useful for customers as is. In fact I think it's harmful as it produces alert fatigue.

Reply:
I think we want the user to actually be alerted when an cluster operator is degraded for that long.

And can you explain why is there a requirement for a runbook for each alert?

Reply:
do we have a better suggestion for the severity?? any guidelines that we can follow?

Reply:
> In fact I think it's harmful as it produces alert fatigue.

I'd rather avoid alert fatigue on these by not hitting these thresholds, not by not monitoring the thresholds :).  All of these things seem like pretty bad "I'd want to know first thing in the morning if this was happening" things.

Reply:
Critical severity is for when you page someone and there is something concrete must be fixed ASAP. I would suggest severity warning for something like this, but it's still too vague of an alert that I would send it to anyone, as it's unclear what to do on this alert.

Even if against our instinct, removing/not using flaky alerts is better than causing alert fatigue, alert fatigue causes people to not trust in our alerts and ignore all of them.

Reply:
The point of this alert is to wake up an engineer because they are broken.  We are increasingly targeting alerts towards our engineering / monitoring team, not just the customers.

Alert fatigue is real but this is a critical failure of a core operator, and we have to react to it.

Reply:
> We are increasingly targeting alerts towards our engineering / monitoring team, not just the customers.

Then this should be an alert evaluated by telemetry infrastructure, not in-cluster.

Reply:
I disagree - we will have customers without telemetry.

Our job is to fix bugs rapidly.  Alerts can be generated by both bugs, user error, and unanticipated states.

It would be irresponsible of us not to alert these conditions on cluster for all disconnected users.